internal/rbd/nodeserver.go

```diff
@@ -216,23 +216,21 @@ func populateRbdVol(
 		rv.RbdImageName = imageAttributes.ImageName
 	}
 
-	krbdSupported := false
-	if req.GetVolumeContext()["mounter"] != rbdNbdMounter {
-		krbdSupported = isKrbdFeatureSupported(ctx, req.GetVolumeContext()["imageFeatures"])
-		if !krbdSupported && !parseBoolOption(ctx, req.GetVolumeContext(), tryOtherMounters, false) {
+	if req.GetVolumeContext()["mounter"] == rbdDefaultMounter &&
+		!isKrbdFeatureSupported(ctx, req.GetVolumeContext()["imageFeatures"]) {
+		if !parseBoolOption(ctx, req.GetVolumeContext(), tryOtherMounters, false) {
 			log.ErrorLog(ctx, "unsupported krbd Feature, set `tryOtherMounters:true` or fix krbd driver")
 
 			return nil, status.Errorf(codes.Internal, "unsupported krbd Feature")
+		} else {
+			// fallback to rbd-nbd,
+			// ignore the mapOptions and unmapOptions as they are meant for krbd use.
+			rv.Mounter = rbdNbdMounter
 		}
-	}
-	if krbdSupported {
+	} else {
+		rv.Mounter = req.GetVolumeContext()["mounter"]
```

Comment:
what if the user has specified generic mount options which are applicable for both krbd/rbd-nbd?

Reply:
Most of the time it would be specific to the mounter type. 

> what if the user has specified generic mount options which are applicable for both krbd/rbd-nbd?

This is why I recommended having mounter-specific mapOptions in the storage class. I don't think we have a way to detect which options belong to which mounter etc, so that we can have one `mapOption`, and adding such a mechanism might lead to additional maintenance hickups.

Reply:
okay. we need to identify a way to achieve this one possible by avoiding adding more SC parameters.

Reply:
Yes! let's take that in a follow-up PR to avoid additional delay with this PR, considering this PR is on the hot seat now :-)

Reply:
@pkalever  @Madhu-1  one idea to avoid having more sc parameters are say you specify mounter in the sc, at that time itself specify the mount options 

Like 

`mounter: "nbd:<nbdmapoptions>"` 

or 

`mounter: "krbd:<krbdmapoptions>"`

Reply:
Good Idea! Humble. Let's see what others got.

Reply:
AFAIK this doesn't solve the problem for the tryOthermonters. IMO this can be solved by adding  `mapOption:"kbrd:v1,v2,v2,nbd:v1,v2,v3" same applies for unmap also?

Reply:
Yes! So its the user's responsibility to tweak mapOptions if he is setting `tryOthermonters: true` in a way that it works for krbd, nbd or some other mounter in the future.

`mapOption: "common: op1, op2 :: krbd-specific: op1, op2 :: nbd-specific: op1, op2"`

delimiters are fancy, can be changed to anything else as per everyone's agreement.

Reply:
Yes, the main problem is to have a proper delimiter for identification

Reply:
yeah, individual mounters can take `mapoptions` as mentioned above, for tryOtherMounters, it can be clubbed as in above..

Reply:
Its horrible, very user unfriendly... but on the other hand, I also do not like a need for mounter-specific configuration options.

So, I'd like to go with:

```yaml
mapOption: "kbrd:v1,v2,v3;nbd:v1,v2,v3"
```

- By omitting `krbd:` or `nbd:`, the option(s) apply to both.
- A user can _override_ the options for a mounter by specifying `krbd:` or `nbd:`. 
  ```yaml
  mapOption: "v1,v2,v3;nbd:v1,v2,v3"
  ```
  is effectively the same as the 1st example.
- Sections are split by `;`.
- No need to specify `common`, keep things as simple as possible. If users want to specify options for both `krbd` and `nbd`, they can mention them twice them.

Reply:
`mapOption` can be kept for common mapping options between all mounters we support, regardless,  I am fine with any of the above models, the main point was that, `using mapping options followed with a delimter to reduce sc params`..

Reply:
@nixpanic well summarized :+1:

Reply:
> By omitting krbd: or nbd:, the option(s) apply to both.

@nixpanic won't this be backward-incompatible ? IMHO this should be treated as options to `rbdDefaultMounter` like before.

Reply:
Here is the issue to track this item https://github.com/ceph/ceph-csi/issues/2641
Please use the issue for any further discussions on this topic.